ACM-11562:writing process from upstream new ACM-MCE docs for discover function (#6629)
open-cluster-management/rhacm-docs: clusters/acm_integration/acm_discover_hosted.adoc

*Required access:* Cluster administrator

* <<hosted-acm-prereqs,Prerequisites>>
* <<hosted-import-config,Configuring {product-title-short} to import {mce-short} clusters>>
* <<hosted-import-mce,Importing {mce-short} manually>>
* <<hosted-discover,Discovering hosted clusters from {mce-short}>>

[#hosted-acm-prereqs]
== Prerequisites

* You need one or more {mce-short} clusters.

* You need a {product-title-short} cluster set as your hub cluster.

* Install the `clusteradm` CLI by running the following command:

+
[source,bash]
----
curl -L https://raw.githubusercontent.com/open-cluster-management-io/clusteradm/main/install.sh | bash
----

[#hosted-import-config]
== Configuring {product-title-short} to import {mce-short} clusters

{mce-short} has a `local-cluster`, which is a hub cluster that is managed. The following default `addons` are enabled for this `local-cluster` in the `open-cluster-management-agent-addon` namespace:

- `cluster-proxy`
- `managed-serviceaccount`
- `work-manager`

[#config-addons-hosted]
=== Configuring add-ons 

When your {mce-short} is imported into {product-title-short}, {product-title-short} enables the same set of add-ons to manage the {mce-short}. 

> Install those add-ons in a different {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` add-ons while  {product-title-short} manages {mce-short} at the same time. Complete the following procedure:

Reviewer: ```suggestion
Install those add-ons in a different {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` add-ons while {product-title-short} manages {mce-short} at the same time. Complete the following procedure:
```

. Log in to your {product-title-short} with the CLI.

> . Create the `addonsDeploymentConfig` resource to specify a different add-on installation namespace. See the following example where `agentInstallNamespace` points to `open-cluster-management-agent-addons-discovery`:

Reviewer 2: `addonsDeploymentConfig` -> `AddOnDeploymentConfig `

`open-cluster-management-agent-addons-discovery` -> `open-cluster-management-agent-addon-discovery`

+
[source,yaml]
----

> apiVersion: addons.open-cluster-management.io/v1alpha1

Reviewer 2: ```
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: AddOnDeploymentConfig
```

kind: addonsDeploymentConfig
metadata:

>   name: addons-ns-config

Reviewer 2: `name: addons-ns-config` -> `name: addon-ns-config`

  namespace: multicluster-engine
spec:

>   agentInstallNamespace: open-cluster-management-agent-addons-discovery

Reviewer 2: `  agentInstallNamespace: open-cluster-management-agent-addons-discovery` -> `  agentInstallNamespace: open-cluster-management-agent-addon-discovery`

----

. Run `oc apply -f <filename>.yaml` to apply the file.

> . Update the existing `ClusterManagementaddons` resources for the add-ons so that the add-ons are installed in the `open-cluster-management-agent-addons-discovery` namespace that is specified in the `addonsDeploymentConfig` resource that you created. See the following example with `open-cluster-management-global-set` as the namespace:

Reviewer 2: `ClusterManagementaddons` -> `ClusterManagementAddOn `
`open-cluster-management-agent-addons-discovery` -> `open-cluster-management-agent-addon-discovery`
`addonsDeploymentConfig ` -> `AddOnDeploymentConfig`

+
[source,yaml]
----

> apiVersion: addons.open-cluster-management.io/v1alpha1

Reviewer 2: `addons` -> `addon`

> kind: ClusterManagementaddons

Reviewer 2: `ClusterManagementaddons` -> `ClusterManagementAddOn`

metadata:
  name: work-manager
spec:

>   addonsMeta:

Reviewer 2: `addonsMeta` -> addonMeta

    displayName: work-manager
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
    type: Placements
----

> .. Add the `addonsDeploymentConfigs` to the `ClusterManagementaddons`. See the following example:

Reviewer: ```suggestion
.. Add the parameter value, `addonsDeploymentConfigs` to the `ClusterManagementaddons` resource. See the following example:
```

Author: I usually eliminate the need for extra words and commas in these in peer reviews. So:

Add the parameter value, `addonsDeploymentConfigs` to the `ClusterManagementaddons` resource. See the following example:

becomes

Add `addonsDeploymentConfigs` to the `ClusterManagementaddons` specification. See the following example:

or 

Add the `addonsDeploymentConfigs` value to the `ClusterManagementaddons` specification. See the following example:

This is for conciseness, and because I often see misuse of commas when I see this. (Here we would have to put one after the value as well, so it ends up being too many commas.)

Author: I will add one of those examples.

Reviewer: That makes sense. Thanks for the reminder. I like the second option

> .. Add the `addonsDeploymentConfigs` to the `ClusterManagementaddons`. See the following example:

Reviewer 2: `addonsDeploymentConfigs` -> `AddOnDeploymentConfig`
`ClusterManagementaddons` -> `ClusterManagementAddOn`

+
[source,yaml]
----

> apiVersion: addons.open-cluster-management.io/v1alpha1

Reviewer 2: ```
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ClusterManagementAddOn
metadata:
  name: work-manager
spec:
  addOnMeta:
    displayName: work-manager
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addon.open-cluster-management.io
        name: addon-ns-config
        namespace: multicluster-engine
        resource: addondeploymentconfigs
    type: Placements
```

addons -> remove `s`

kind: ClusterManagementaddons
metadata:
  name: work-manager
spec:
  addonsMeta:
    displayName: work-manager
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io
        name: addons-ns-config
        namespace: multicluster-engine
        resource: addonsdeploymentconfigs
    type: Placements
----

> .. Add the `addonsDeploymentConfig` to the `managed-serviceaccount`. See the following example:

Reviewer: ```suggestion
.. Add the `addonsdeploymentconfigs` value to the `ClusterManagementaddons` specification named `managed-serviceaccount`. See the following example:
```

Author: same here

> .. Add the `addonsDeploymentConfig` to the `managed-serviceaccount`. See the following example:

Reviewer 2: `addonsDeploymentConfig` -> `AddOnDeploymentConfig`

+
[source,yaml]
----

> apiVersion: addons.open-cluster-management.io/v1alpha1

Reviewer 2: ```
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ClusterManagementAddOn
metadata:
  name: managed-serviceaccount
spec:
  addOnMeta:
    displayName: managed-serviceaccount
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addon.open-cluster-management.io
        name: addon-ns-config
        namespace: multicluster-engine
        resource: addondeploymentconfigs
    type: Placements
```

Replace the whole YAML, do not change addon -> addons

kind: ClusterManagementaddons
metadata:
  name: managed-serviceaccount
spec:
  addonsMeta:
    displayName: managed-serviceaccount
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io
        name: addons-ns-config
        namespace: multicluster-engine

>         resource: addonsdeploymentconfigs

Author: @dockerymick we cannot ever change the YAML that is give to us without talking to development because this is used. If not the same then the content changes before the code or we check with engineering.

Reviewer: Understood, I updated my other suggestion. Please ignore this

    type: Placements
----

> .. Add the `addonsDeploymentConfig` to the `cluster-proxy` add-ons. See the following example:

Reviewer: ```suggestion
.. Add the `addonsdeploymentconfigs` value to the `ClusterManagementaddons` resource named, `cluster-proxy`. See the following example:
```

Author: same here.

Reviewer: Updated the suggestion

Author: Perfect! Looks great.

> .. Add the `addonsDeploymentConfig` to the `cluster-proxy` add-ons. See the following example:

Reviewer 2: addonsDeploymentConfig -> addonDeploymentConfig

+
[source,yaml]
----

> apiVersion: addons.open-cluster-management.io/v1alpha1

Reviewer 2: addons.open-cluster-management.io/v1alpha1 -> addon.open-cluster-management.io/v1alpha1

> kind: ClusterManagementaddons

Reviewer 2: ClusterManagementaddons -> ClusterManagementAddOn

metadata:
  name: cluster-proxy
spec:

>   addonsMeta:

Reviewer 2: addonsMeta -> addonMeta

    displayName: cluster-proxy
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io

>         name: addons-ns-config

Reviewer 2: addons-ns-config -> addon-ns-config

Reviewer 2: Just replace the whole YAML with this and do not change anything in the YAML

```
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ClusterManagementAddOn
metadata:
  name: cluster-proxy
spec:
  addOnMeta:
    displayName: cluster-proxy
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addon.open-cluster-management.io
        name: addon-ns-config
        namespace: multicluster-engine
        resource: addondeploymentconfigs
    type: Placements
```

        namespace: multicluster-engine

>         resource: addonsdeploymentconfigs

Author: same

    type: Placements
----

> . Run the following command to verify that the add-ons for the {product-title-short} `local-cluster` are re-installed into the namespace that you specified.:

Reviewer: Extra period
```suggestion
. Run the following command to verify that the add-ons for the {product-title-short} `local-cluster` are re-installed into the namespace that you specified:
```

+
[source,bash]
----

> oc get deployment -n open-cluster-management-agent-addons-discovery

Reviewer 2: open-cluster-management-agent-addons-discovery -> open-cluster-management-agent-addon-discovery

----

+
See the following output example:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h

> klusterlet-addons-workmgr             1/1     1            1           24h

Reviewer 2: klusterlet-addons-workmgr -> klusterlet-addon-workmgr

> managed-serviceaccount-addons-agent   1/1     1            1           24h

Reviewer 2: managed-serviceaccount-addons-agent  -> managed-serviceaccount-addon-agent

> ----

Reviewer: Slight spacing issue

<img width="965" alt="Screenshot 2024-07-15 at 3 05 54 PM" src="https://github.com/user-attachments/assets/79a08ad4-b682-414e-a3a5-15ecfd6ee11d">

```suggestion
See the following output example:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE    AGE
cluster-proxy-proxy-agent             1/1     1            1           24h
klusterlet-addons-workmgr             1/1     1            1           24h
managed-serviceaccount-addons-agent   1/1     1            1           24h
----
```

Author: but this is not grayed out like it should be so that is another issue here

[#create-klusterletconfig-mce]
=== Creating a _KlusterletConfig_ resource

{mce-short} has a local-cluster, which is a hub cluster that is managed. A resource named `klusterlet` is created for this local-cluster.

> When your {mce-short} is imported into {product-title-short}, {product-title-short} installs the klusterlet with the same name, `klusterlet`, to manage the {mce-short}, which conflicts with the {mce-short} local-cluster. klusterlet.

Reviewer: Long statement 
```suggestion
When your {mce-short} is imported into {product-title-short}, {product-title-short} installs the klusterlet with the same name, `klusterlet`, to manage the {mce-short}. This conflicts with the {mce-short} local-cluster klusterlet.
```

You need to create a `KlusterletConfig` resource that is used by `ManagedCluster` resources to import {mce-short} clusters so that the klusterlet is installed with a different name to avoid the conflict. Complete the following procedure:

> . Create a `KlusterletConfig` using the following example. When this `KlusterletConfig` is referenced in a `ManagedCluster`, the value in the `spec.installMode.noOperator.postfix` field is used as a suffix to the klusterlet name, such as `klusterlet-mce-import`:

Reviewer: ```suggestion
. Create a `KlusterletConfig` resource using the following example. When this `KlusterletConfig` resource is referenced in a managed cluster, the value in the `spec.installMode.noOperator.postfix` field is used as a suffix to the klusterlet name, such as `klusterlet-mce-import`:
```

+
[source,yaml]
----
kind: KlusterletConfig
apiVersion: config.open-cluster-management.io/v1alpha1
metadata:
  name: mce-import-klusterlet-config
spec:
  installMode:
    type: noOperator
    noOperator:
       postfix: mce-import
----

. Run `oc apply -f <filename>.yaml` to apply the file.

[#backup-hosted-acm]
=== Configure for backup and restore

> Since you installed {product-title-short} , you can also use the _Backup and restore_ feature.

Reviewer: ```suggestion
Since you installed {product-title-short}, you can also use the _Backup and restore_ feature.
```

If the hub cluster is restored in a disaster recovery scenario, the imported {mce-short} clusters and hosted clusters are imported to the newer {product-title-short} hub cluster. 

In this scenario, you need to restore the previous configurations as part of {product-title-short} hub cluster restore. 

> Add `backup=true` to enable backup. See the following steps for each add-on:

Reviewer: ```suggestion
Add the `backup=true` label to enable backup. See the following steps for each add-on:
```

Author: YES!! That is how the ones above should be without the extra need for commas. Great suggestion.

> * For your addons-ns-config`, run the following command:

Reviewer: ```suggestion
* For your `addons-ns-config`, run the following command:
```

> * For your addons-ns-config`, run the following command:

Reviewer 2: addons-ns-config -> addon-ns-config

> * For your `addons-ns-config`, run the following command:

Reviewer 2: addons-ns-config -> addon-ns-config

+
[source,bash]
----

> oc label addonsdeploymentconfig addons-ns-config -n multicluster-engine cluster.open-cluster-management.io/backup=true

Reviewer 2: addonsdeploymentconfig -> addondeploymentconfig (remove s)

addons-ns-config -> addon-ns-config

----

> * For your `hypershift-addons-deploy-config`, run the following command:

Reviewer 2: hypershift-addons-deploy-config -> hypershift-addon-deploy-config

+
[source,bash]
----

> oc label addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine cluster.open-cluster-management.io/backup=true

Reviewer 2: addonsdeploymentconfig -> addondeploymentconfig (remove s)

hypershift-addons-deploy-config -> hypershift-addon-deploy-config

----

* For your `work-manager`, run the following command:

+
[source,bash]
----

> oc label clustermanagementaddons work-manager cluster.open-cluster-management.io/backup=true

Reviewer 2: clustermanagementaddons -> clustermanagementaddon (remove s)

----

* For your `cluster-proxy `, run the following command:

+
[source,bash]
----

> oc label clustermanagementaddons cluster-proxy cluster.open-cluster-management.io/backup=true

Reviewer 2: clustermanagementaddons -> clustermanagementaddon (remove s)

----

* For your `managed-serviceaccount`, run the following command:

+
[source,bash]
----

> oc label clustermanagementaddons managed-serviceaccount cluster.open-cluster-management.io/backup=true

Reviewer 2: clustermanagementaddons -> clustermanagementaddon (remove s)

----

* For your `mce-import-klusterlet-config`, run the following command:

+
[source,bash]
----
oc label KlusterletConfig mce-import-klusterlet-config cluster.open-cluster-management.io/backup=true
----

> 

Reviewer: Is there a verification step?

Reviewer 2: The verification step would be to check the `KlusterletConfig` resource to ensure the label is added.

Reviewer: @rokej thanks, will you also share the command that the user needs to run to see the `KlusterletConfig`?

Reviewer 2: In this entire backup and restore section, we added the backup label to six resources in three different resource kinds. To verify that the backup label is added to all six resources, you can run these commands.

```
oc get addondeploymentconfig -l 'cluster.open-cluster-management.io/backup=true' -A
oc get clustermanagementaddon -l 'cluster.open-cluster-management.io/backup=true' -A
oc get KlusterletConfig -l 'cluster.open-cluster-management.io/backup=true' -A
```

Author: I cannot get this in today.

Reviewer: @swopebe no worries, like you mentioned you can get this is later

Author: I have to open an issue for a diagram anyway.

[#hosted-import-mce]
== Importing {mce-short} manually

> 

Reviewer: Short description needed

Reviewer: ```suggestion

To manually import an {mce-short} cluster from your {product-title-short} cluster, complete the following procedure:
```

> . From your {product-title-short} cluster, create a `ManagedCluster` resource manually to import an {mce-short} cluster. See the following file example:

Reviewer: ```suggestion
.  Create a `ManagedCluster` resource manually to import an {mce-short} cluster. See the following file example:
```

Author: @dockerymick I see you are trying to make this concise. In an ACM topic, you are right to get rid of unnecessary instructions bc the user is in their ACM cluster, but here we have ACM and MCE....it is an integration topic. I think this needs to stay, as it was designed to orient the user.

----
apiVersion: cluster.open-cluster-management.io/v1
kind: ManagedCluster
metadata:
  annotations:
    agent.open-cluster-management.io/klusterlet-config: mce-import-klusterlet-config <1>
  name: mce-a <2>
spec:
  hubAcceptsClient: true
  leaseDurationSeconds: 60
----
<1> The `mce-import-klusterlet-config` annotation references the `KlusterletConfig` resource that you created in the previous step to install the {product-title-short} klusterlet with a different name in {mce-short}.
<2> The example imports an {mce-short} managed cluster named `mce-a`.

. Run `oc apply -f <filename>.yaml` to apply the file.

. Create the `auto-import-secret` secret that references the `kubeconfig` of the {mce-short} cluster. Go to xref:../cluster_lifecycle/import_cli.adoc#importing-clusters-auto-import-secret [Importing a cluster by using the auto import secret] to add the auto import secret to complete the {mce-short} auto-import process. 

+
After you create the auto import secret in the {mce-short} managed cluster namespace in the {product-title-short} cluster, the managed cluster is registered.

. Run the following command to get the status:

+
[source,bash]
----
oc get managedcluster
----

+

> See following example output with the status and example urls of managed clusters:

Reviewer: ```suggestion
See following example output with the status and example URLs of managed clusters:
```

+
[source,bash]
----
NAME           HUB ACCEPTED   MANAGED CLUSTER URLS            JOINED   AVAILABLE   AGE
local-cluster  true           https://<api.acm-hub.com:port>  True     True        44h
mce-a          true           https://<api.mce-a.com:port>    True     True        27s
----

*Important:* Do not enable any other {product-title-short} add-ons for the imported {mce-short}.

[#hosted-discover]
== Discovering hosted clusters

> After all your {mce-short} clusters are imported into {product-title-short}, you need to enable the hypershift add-on for those managed {mce-short} clusters to discover the hosted clusters.

Reviewer: ```suggestion
After all your {mce-short} clusters are imported into {product-title-short}, you need to enable the HyperShift add-on for those managed {mce-short} clusters to discover the hosted clusters.
```

Author: @dockerymick this is intentionally written in proper format. If we do not put it in code as improper, if improper in code, this needs to be written the same as the title. I almost always double check. This is correct:

https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html/hosted_control_planes/hosted-control-planes-overview#hosted-control-planes-benefits_hcp-overview

It could also be coded, but not improper in this case.

Reviewer: Ok, throughout our doc i saw HyperShift

Author: Good to catch these. 

To give you some context:

This was my draft to work off of, but the issue had numerous links to start from and we changed it a lot.

 https://github.com/stolostron/hypershift-addon-operator/blob/main/docs/discovering_hostedclusters.md#auto-importing-the-discovered-hosted-clusters

> Default add-ons are installed in to a different namespace in the previous procedures. Similarly, you install the `hypershift-addons` in to a different namespace in {mce-short} so that the add-ons agent for {mce-short} local-cluster and the agent for {product-title-short} can work in {mce-short}. 

Reviewer: ```suggestion
Default add-ons are installed into a different namespace in the previous procedures. Similarly, you install the `hypershift-addons` into a different namespace in {mce-short} so that the add-ons agent for {mce-short} local-cluster and the agent for {product-title-short} can work in {mce-short}. 
```

Author: great catch

*Important:* For all the following commands, replace `<managed-cluster-names>` with comma-separated managed cluster names for {mce-short}.

> . Run the following command to set the `agentInstallNamespace` namespace of the add-on to `open-cluster-management-agent-addons-discovery`:

Reviewer 2: `open-cluster-management-agent-addons-discovery` -> `open-cluster-management-agent-addon-discovery`

+
[source,bash]
----

> oc patch addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"agentInstallNamespace":"open-cluster-management-agent-addons-discovery"}}'

Reviewer 2: oc patch addondeploymentconfig hypershift-addon-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"agentInstallNamespace":"open-cluster-management-agent-addon-discovery"}}'

> oc patch addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"agentInstallNamespace":"open-cluster-management-agent-addons-discovery"}}'

Reviewer: ```suggestion
Complete the following procedure:

. Run the following command to set the `agentInstallNamespace` namespace of the add-on to `open-cluster-management-agent-addons-discovery`:

+
[source,bash]
----
oc patch addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"agentInstallNamespace":"open-cluster-management-agent-addons-discovery"}}'
```

----

> . Run the following command to disable metrics and to disable the hypershift operator management:

Reviewer: ```suggestion
. Run the following command to disable metrics and to disable the HyperShift operator management:
```

+
[source,bash]
----

> oc patch addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"customizedVariables":[{"name":"disableMetrics","value": "true"},{"name":"disableHOManagement","value": "true"}]}}'

Reviewer 2: oc patch addondeploymentconfig hypershift-addon-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"customizedVariables":[{"name":"disableMetrics","value": "true"},{"name":"disableHOManagement","value": "true"}]}}'

----

. Run the following command to enable the `hypershift-addon` for {mce-short}:

+
[source,bash]
----

> clusteradm addons enable --names hypershift-addon --clusters <managed-cluster-names>

Reviewer 2: clusteradm addon enable --names hypershift-addon --clusters <managed-cluster-names>

----

. You can get the {mce-short} managed cluster names by running the following command in {product-title-short}.

+
[source,bash]
----
oc get managedcluster
----

> . Log into {mce-short} clusters and verify that the hypershift add-on is installed in the namespace that you specified. Run the following command:

Reviewer: ```suggestion
. Log into {mce-short} clusters and verify that the HyperShift add-on is installed in the namespace that you specified. Run the following command:
```

+
[source,bash]
----

> oc get deployment -n open-cluster-management-agent-addons-discovery

Reviewer 2: oc get deployment -n open-cluster-management-agent-addon-discovery

----
 

> See the following example output that lists the add-ons:

Reviewer: fixing alignment
```suggestion
+
See the following example output that lists the add-ons:
```

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h

> klusterlet-addons-workmgr            1/1     1            1           24h

Reviewer 2: klusterlet-addon-workmgr

> hypershift-addons-agent              1/1     1            1           24h

Reviewer 2: hypershift-addon-agent

> managed-serviceaccount-addons-agent  1/1     1            1           24h

Reviewer 2: managed-serviceaccount-addon-agent

----

{product-title-short} deploys the `hypershift-addon`, which is the discovery agent that discovers hosted clusters from {mce-short}. The agent creates the corresponding `DiscoveredCluster` custom resource in the {mce-short} managed cluster namespace in the {product-title-short} hub cluster when the hosted cluster `kube-apiserver` becomes available. 

> You can view your discovered clusters in the console.

Reviewer: Should this be a new section or should it be considered a verification step? If verification, I also recommend that the order steps become substeps

Author: This is not a verification step.

Author: Keeping as-is, I can look at organization again in the future.

. Log into hub cluster console and navigate to *All Clusters* > *Infrastructure* > *Clusters*. 
. Find the _Discovered clusters_ tab to view all discovered hosted clusters from {mce-short} with type `MultiClusterEngineHCP`. 

> Next, visit xref:../acm_integrate_import_hcp.adoc#auto-import-hcp[Automating import for discovered hosted clusters] to learn how to automatically import clusters.

Reviewer: Is this a part of the steps? If not, I recommend that this be added to an Additional resources section in this file

Reviewer: Also this link is broken 
```suggestion
Next, visit xref:../acm_integration/acm_integrate_import_hcp.adoc#auto-import-hcp[Automating import for discovered hosted clusters] to learn how to automatically import clusters.
```

Author: @dockerymick this change I have a question about. Usually if we are in the same book, which we are, we don't need the extra path, we just go back into the folder (../) and down to the file. (?)

Reviewer: Yes, I agree but it appeared as broken when I ran the tool 
<img width="1126" alt="Screenshot 2024-07-16 at 9 40 20 AM" src="https://github.com/user-attachments/assets/9b8093b0-6b98-447b-ba18-75407febb249">

Author: @dockerymick Ok. I think it is best to understand _why_ when these things pop up, that is why I asked. Seems we don't have an answer right now.

*Note:* Since the hosted control planes run on the managed {mce-short} cluster nodes, the number of hosted control planes that the cluster can host is determined by the resource availability of managed {mce-short} cluster nodes, as well as the number of managed {mce-short} clusters.You can add more nodes or managed clusters to host more hosted control planes.

*Required access:* Cluster administrator

* <<discover-prereqs, Prerequisites>>
* <<config-acm-import, Configuring {product-title-short} to import {mce-short} clusters>>
* <<import-mce-manually,Importing {mce-short} manually>>
* <<discover-hosted-acm-mce, Discovering hosted clusters from {mce-short}>>

[#discover-hosted-acm-prereqs]
== Prerequisites

* You need one or more {mce-short} clusters.

* You need a {product-title-short} cluster set as your hub cluster.

* Install the `clusteradm` CLI by running the following command:

+
[source,bash]
----
curl -L https://raw.githubusercontent.com/open-cluster-management-io/clusteradm/main/install.sh | bash
----

[#config-acm-import-hosted]
== Configuring {product-title-short} to import {mce-short} clusters

> {mce-short} has a `local-cluster`, which is a hub cluster that is managed. The following default `addons` are enabled for this `local-cluster` in the `open-cluster-management-agent-addon` namespace:

Reviewer 2: Move line 35 ~ 39 to the beginning of `Configuring addons` section (in line 43).

Author: @rokej If I do that, that leaves no text between headers and we cannot do that for accessibility reasons, we have to have content between headers. Willing to take suggestions tho.

Reviewer 2: @swopebe OK. In that case, let's add one more line to give the reader a little more context. We can say..

This section describes how to configure these addons so that they are installed in a different namespace prior to importing MCEs.

        name: addons-ns-config
        namespace: multicluster-engine
        resource: addonsdeploymentconfigs
    type: Placements
----

. Run the following command to verify that the addons for the {product-title-short} `local-cluster` are re-installed into the namespace that you specified.:

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-addons-discovery
----

+
See the following output example:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h
klusterlet-addons-workmgr             1/1     1            1           24h
managed-serviceaccount-addons-agent   1/1     1            1           24h
----

[#create-klusterletconfig]
=== Creating a _KlusterletConfig_ resource

{mce-short} has a local-cluster, which is a hub cluster that is managed. A resource named `klusterlet` is created for this local-cluster.

When your {mce-short} is imported into {product-title-short}, {product-title-short} installs the klusterlet with the same name, `klusterlet`, to manage the {mce-short}, which conflicts with the {mce-short} local-cluster. klusterlet.

> Install those addons in a _different_ {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` addons, while {product-title-short} manages {mce-short} at the same time. 

Reviewer 2: Remove this line.

+
[source,bash]
----
oc label addondeploymentconfig addon-ns-config -n multicluster-engine cluster.open-cluster-management.io/backup=true
----

* For your `hypershift-addon-deploy-config`, run the following command:

+
[source,bash]
----
oc label addondeploymentconfig hypershift-addon-deploy-config -n multicluster-engine cluster.open-cluster-management.io/backup=true
----

* For your `work-manager`, run the following command:

+
[source,bash]
----
oc label clustermanagementaddon work-manager cluster.open-cluster-management.io/backup=true
----

* For your `cluster-proxy `, run the following command:

+
[source,bash]
----
oc label clustermanagementaddon cluster-proxy cluster.open-cluster-management.io/backup=true
----

* For your `managed-serviceaccount`, run the following command:

+
[source,bash]
----
oc label clustermanagementaddon managed-serviceaccount cluster.open-cluster-management.io/backup=true
----

* For your `mce-import-klusterlet-config`, run the following command:

+
[source,bash]
----
oc label KlusterletConfig mce-import-klusterlet-config cluster.open-cluster-management.io/backup=true
----

[#hosted-import-mce]
== Importing {mce-short} manually

> To manually import an {mce-short} cluster from your {product-title-short} cluster, complete the following procedure:
> . From your {product-title-short} cluster, create a `ManagedCluster` resource manually to import an {mce-short} cluster. See the following file example:
> 

Reviewer: ```suggestion
To manually import an {mce-short} cluster from your {product-title-short} cluster, complete the following procedure:

. From your {product-title-short} cluster, create a `ManagedCluster` resource manually to import an {mce-short} cluster. See the following file example:

```

+
[source,bash]
----
oc patch addondeploymentconfig hypershift-addon-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"agentInstallNamespace":"open-cluster-management-agent-addon-discovery"}}'
----

. Run the following command to disable metrics and to disable the HyperShift operator management:

+
[source,bash]
----
oc patch addondeploymentconfig hypershift-addon-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"customizedVariables":[{"name":"disableMetrics","value": "true"},{"name":"disableHOManagement","value": "true"}]}}'
----

. Run the following command to enable the `hypershift-addon` for {mce-short}:

+
[source,bash]
----
clusteradm addon enable --names hypershift-addon --clusters <managed-cluster-names>
----

. You can get the {mce-short} managed cluster names by running the following command in {product-title-short}.

+
[source,bash]
----
oc get managedcluster
----

. Log into {mce-short} clusters and verify that the `hypershift-addon` is installed in the namespace that you specified. Run the following command:

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-addon-discovery
----

>  
> See the following example output that lists the add-ons:
> 
> +
> [source,bash]
> ----
> NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
> cluster-proxy-proxy-agent            1/1     1            1           24h
> klusterlet-addon-workmgr            1/1     1            1           24h
> hypershift-addon-agent              1/1     1            1           24h
> managed-serviceaccount-addon-agent  1/1     1            1           24h
> ----

Reviewer: ```suggestion
 +
See the following example output that lists the add-ons:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h
klusterlet-addon-workmgr            1/1     1            1           24h
hypershift-addon-agent              1/1     1            1           24h
managed-serviceaccount-addon-agent  1/1     1            1           24h
----
```

[#discovering-hcp-acm]
= Discovering {mce-short} hosted clusters in {product-title-short}

If you have {mce-short} clusters that are hosting many hosted clusters, you can bring these hosted clusters to a {product-title-short} hub cluster to manage with {product-title-short} management components, such as _Application lifecycle_ and _Governance_.

You can have those hosted clusters automatically discovered and imported as managed clusters.

Since the hosted control planes run on the managed {mce-short} cluster nodes, the number of hosted control planes the cluster can host is determined by the resource availability of managed {mce-short} cluster nodes, as well as the number of managed {mce-short} clusters. 

You can add more nodes or managed clusters to host more hosted control planes.

*Required access:* Cluster administrator

* <<-manage-discover-prereqs,Prerequisites>>
* Configuring {product-title-short} to import {mce-short} clusters
* Importing {mce-short} 
* Discovering hosted clusters from MCE
* This requires `oc` and `clusteradm` CLIs.

> //find this link

Author: Wed: line 18 we said we would get a link.

Reviewer 2: Here is the actual command to install `clusteradm` . We were going to remove `oc` from the pre-req.

```
curl -L https://raw.githubusercontent.com/open-cluster-management-io/clusteradm/main/install.sh | bash
```

[#discover-hosted-acm-prereqs]
== Prerequisites

* You need one or more {mce-short} hosting clusters that are _managed_ clusters. These clusters are not {ocp} or {product-title-short} clusters because {mce-short} installs other operators, such as Hive and BareMetal infrastructure operators that you need.

* You need a {product-title-short} cluster set as your hub cluster.

[#config-acm-import-hosted]
== Configuring {product-title-short} to import hosted clusters

{mce-short} has a `local-cluster`, which is a hub cluster that is managed. The default add-ons are enabled for this `local-cluster`. Complete the following procedure to configure your local cluster to import:

. Run the following command to get the list of `managedclusteradd-ons` for your `local-cluster`:

+
[source,bash]
----
oc get managedclusteradd-ons -n local-cluster
----

See the example output of available add-ons:

+
[source,bash]
----
NAME                     AVAILABLE   DEGRADED   PROGRESSING
cluster-proxy            True                   False
hypershift-add-ons       True        False      False
managed-serviceaccount   True                   False
work-manager             True                   False
----

> . Run the following command to get your add-ons:

Author: Line 52 and line 61, let's explain what they are doing and looking for?

Reviewer 2: This is for checking that the add ons are installed in the default namespace `open-cluster-management-agent-addons` (the namespace in line 58 is wrong. addons, not add-ons) before making changes to the addon installation namespace configuration.

hypershift-add-ons-agent               1/1     1            1         25h
klusterlet-add-ons-workmgr             1/1     1            1         25h
managed-serviceaccount-add-ons-agent   1/1     1            1         25h
----

[#config-add-ons-mce]
=== Configuring add-ons 

When your {mce-short} is imported into {product-title-short}, {product-title-short} enables the same set of add-ons to manage the {mce-short}. 

Install those add-ons in a different {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` add-ons while  {product-title-short} manages {mce-short} at the same time. Complete the following procedure:

. Log in to your {product-title-short} with the CLI.

. Create the `add-onsDeploymentConfig` resource to specify a different add-on installation namespace. See the following example where `agentInstallNamespace` points to `open-cluster-management-agent-add-ons-discovery`:

+
[source,yaml]
----
apiVersion: add-ons.open-cluster-management.io/v1alpha1
kind: add-onsDeploymentConfig
metadata:
  name: add-ons-ns-config
  namespace: multicluster-engine
spec:
  agentInstallNamespace: open-cluster-management-agent-add-ons-discovery
----

> . Update the existing `ClusterManagementadd-ons` resources for the add-ons so that the `add-onsDeploymentConfig` resource have the `agentInstallNamespace` namespace that you created.

Author: please look at this little section?

Author: comment on line 98

Reviewer 2: This is still a part of the first high level step of configuring ACM to import MCE.

----
apiVersion: add-ons.open-cluster-management.io/v1alpha1
kind: ClusterManagementadd-ons
metadata:
  name: managed-serviceaccount
spec:
  add-onsMeta:
    displayName: managed-serviceaccount
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: add-ons.open-cluster-management.io
        name: add-ons-ns-config
        namespace: multicluster-engine
        resource: add-onsdeploymentconfigs
    type: Placements
----

. Add the `add-onsDeploymentConfig` to the `cluster-proxy` add-ons. See the following example:

+
[source,yaml]
----
apiVersion: add-ons.open-cluster-management.io/v1alpha1
kind: ClusterManagementadd-ons
metadata:
  name: cluster-proxy
spec:
  add-onsMeta:
    displayName: cluster-proxy
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: add-ons.open-cluster-management.io
        name: add-ons-ns-config
        namespace: multicluster-engine
        resource: add-onsdeploymentconfigs
    type: Placements
----

> //we need to apply the file.

Author: apply file command, right?

Reviewer 2: No applying file. Line 173 says add the `addondeploymentconfig` section to the existing resource. Also I noticed that you replaced all `addons` with `add-ons`. In line 195, it should be `addondeploymentconfigs`

Author: That was to catch the text, I forgot to exclude the code, that is fixed now.

The add-ons for the {product-title-short} `local-cluster` and all other managed clusters are re-installed into the namespace that you specified. 

Run the following command to verify:

> //??

Author: Is this a verification step, mine 203?

Reviewer 2: Yes it is a verification step.

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-add-ons-discovery
----

. See the following output:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h
klusterlet-add-ons-workmgr             1/1     1            1           24h
managed-serviceaccount-add-ons-agent   1/1     1            1           24h
----

[#create-klusterletconfig]
=== Creating a _KlusterletConfig_ resource

When a `ManagedCluster` references the `KlusterletConfig` resource, the managed cluster `klusterlet` is installed in the namespace that you specified in the `KlusterletConfig`. 

Create a `KlusterletConfig` resource that is used by `ManagedCluster` resources to import {mce-short} clusters. 

You can import the {product-title-short} klusterlet to be installed in a different namespace than the {mce-short} klusterlet for the `local-cluster` in the {mce-short} cluster.

> //still struggling with this

Author: is this clear, you think?

Reviewer 2: Lines 226 ~ 230, I want to say...

When you install an MCE, the MCE hub cluster imports itself as `local-cluster` to be self managed and it installs a klusterlet named `kusterlet` for `local-cluster`. When you import an MCE cluster, ACM installs a klusterlet named `klusterlet` in the MCE cluster by default and it will conflict with the existing MCE local-cluster klusterlet. 

Create a `KlusterletConfig` resource that is used by the `ManagedCluster` resource to import {mce-short} clusters so that ACM installs the klusterlet with a different name. When the following `KlusterletConfig` is used, the klusterlet will be named `klusterlet-mce-import`.

Author: I will come back to this Friday.

> . Create a `KlusterletConfig` using the following example:

Author: Is this the correct way to describe this step? line 233 and after?

Reviewer 2: yes

//?

+
[source,yaml]
----
kind: KlusterletConfig
apiVersion: config.open-cluster-management.io/v1alpha1
metadata:
  name: mce-import-klusterlet-config
spec:
  installMode:
    type: noOperator
    noOperator:
       postfix: mce-import
----

> //apply?

Author: apply?

Reviewer 2: Yes apply.

```
$ cat <<EOF | oc apply -f -
kind: KlusterletConfig
apiVersion: config.open-cluster-management.io/v1alpha1
metadata:
  name: mce-import-klusterlet-config
spec:
  installMode:
    type: noOperator
    noOperator:
       postfix: mce-import
EOF
```

Author: ok, will catch this up. TY

Author: @rokej we would add the command and the YAML separately to the doc, taking out `$ cat <<EOF`, do you mind showing what that should look like?

Reviewer 2: Taking out the `cat` command, it looks like this which is what you already have up there.

```
kind: KlusterletConfig
apiVersion: config.open-cluster-management.io/v1alpha1
metadata:
  name: mce-import-klusterlet-config
spec:
  installMode:
    type: noOperator
    noOperator:
       postfix: mce-import
```

+
[source,bash]
----
oc label clustermanagementadd-ons work-manager cluster.open-cluster-management.io/backup=true
----

* For your `cluster-proxy `, run the following command:

+
[source,bash]
----
oc label clustermanagementadd-ons cluster-proxy cluster.open-cluster-management.io/backup=true
----

* For your `managed-serviceaccount`, run the following command:

+
[source,bash]
----
oc label clustermanagementadd-ons managed-serviceaccount cluster.open-cluster-management.io/backup=true
----

* For your `mce-import-klusterlet-config`, run the following command:

+
[source,bash]
----
oc label KlusterletConfig mce-import-klusterlet-config cluster.open-cluster-management.io/backup=true
----

[#import]
== Importing {mce-short} manually

. From your {product-title-short} cluster, create a `ManagedCluster` resource manually to import an {mce-short} cluster. 

+
[source,yaml]
----
apiVersion: cluster.open-cluster-management.io/v1
kind: ManagedCluster
metadata:
  annotations:
    agent.open-cluster-management.io/klusterlet-config: mce-import-klusterlet-config <1>
  name: mce-a <2>
spec:
  hubAcceptsClient: true
  leaseDurationSeconds: 60
----

> 

Author: I used callouts for this, please check?

Reviewer 2: Looks good

1. The `mce-import-klusterlet-config` annotation references the `KlusterletConfig` resource that you created in the previous step to install the {product-title-short} klusterlet into a different namespace in {mce-short}.
2. The example imports an {mce-short} managed cluster named `mce-a`.

The managed cluster and the namespace is created in the {product-title-short} cluster. 
//verify

> ????

Author: Verification?

You can have those hosted clusters automatically discovered and imported as managed clusters.

*Note:* Since the hosted control planes run on the managed {mce-short} cluster nodes, the number of hosted control planes that the cluster can host is determined by the resource availability of managed {mce-short} cluster nodes, as well as the number of managed {mce-short} clusters.You can add more nodes or managed clusters to host more hosted control planes.

*Required access:* Cluster administrator

* <<discover-prereqs, Prerequisites>>
* <<config-acm-import, Configuring {product-title-short} to import {mce-short} clusters>>
* <<import-mce-manually,Importing {mce-short} manually>>
* <<discover-hosted-acm-mce, Discovering hosted clusters from {mce-short}>>

[#discover-hosted-acm-prereqs]
== Prerequisites

* You need one or more {mce-short} hosting clusters that are _managed_ clusters. These clusters are not {ocp} or {product-title-short} clusters because {mce-short} installs other operators, such as Hive and BareMetal infrastructure operators that you need.

* You need a {product-title-short} cluster set as your hub cluster.

> * You need the `clusteradm` CLI in addition to `oc`.

Author: We need a link for that, yes?

[#discovering-hcp-acm]
= Discovering {mce-short} hosted clusters in {product-title-short}

If you have {mce-short} clusters that are hosting multiple _hosted clusters_, you can bring those hosted clusters to a {product-title-short} hub cluster to manage with {product-title-short} management components, such as _Application lifecycle_ and _Governance_.

You can have those hosted clusters automatically discovered and imported as managed clusters.

*Note:* Since the hosted control planes run on the managed {mce-short} cluster nodes, the number of hosted control planes that the cluster can host is determined by the resource availability of managed {mce-short} cluster nodes, as well as the number of managed {mce-short} clusters.You can add more nodes or managed clusters to host more hosted control planes.

*Required access:* Cluster administrator

* <<discover-prereqs, Prerequisites>>
* <<config-acm-import, Configuring {product-title-short} to import {mce-short} clusters>>
* <<import-mce-manually,Importing {mce-short} manually>>
* <<discover-hosted-acm-mce, Discovering hosted clusters from {mce-short}>>

[#discover-hosted-acm-prereqs]
== Prerequisites

> * You need one or more {mce-short} hosting clusters that are _managed_ clusters. These clusters are not {ocp} or {product-title-short} clusters because {mce-short} installs other operators, such as Hive and BareMetal infrastructure operators that you need.

Reviewer 2: You need one or more {mce-short} clusters.   (Let's keep it simple)

* You need a {product-title-short} cluster set as your hub cluster.

+
[source,bash]
----

> curl -L https://raw.githubusercontent.com/open-cluster-management-io/clusteradm/main/install.sh | bash

Reviewer 2: We need a description for this command. "Install the clusteradm CLI by running the following command.".

Author: I must have lost that in the copy/paste, it was there lol

----

[#config-acm-import-hosted]

> == Configuring {product-title-short} to import hosted clusters

Reviewer 2: Configuring {product-title-short} to import MCE clusters

> {mce-short} has a `local-cluster`, which is a hub cluster that is managed. The default addons are enabled for this `local-cluster`. Complete the following procedure to configure your local cluster to import:

Reviewer 2: Let's remove line 32 ~ 74. This could be confusing. I will add a summary of all these lines at the beginning of the next section.

. Run the following command to get the list of `managedclusteraddons` for your `local-cluster`:

+
[source,bash]
----
oc get managedclusteraddons -n local-cluster
----

See the example output of available addons:

+
[source,bash]
----
NAME                     AVAILABLE   DEGRADED   PROGRESSING
cluster-proxy            True                   False
hypershift-addons       True        False      False
managed-serviceaccount   True                   False
work-manager             True                   False
----

. Run the following command to get your addons:

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-addons
----

See the following output that lists the agent addons:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent             1/1     1            1         25h
hypershift-addons-agent               1/1     1            1         25h
klusterlet-addons-workmgr             1/1     1            1         25h
managed-serviceaccount-addons-agent   1/1     1            1         25h
----

[#config-addons-mce]
=== Configuring addons 

> 

Reviewer 2: Add this at the beginning of this section.

{mce-short} has a `local-cluster`, which is a hub cluster that is managed. The following default addons are enabled for this `local-cluster` in `open-cluster-management-agent-addon` namespace.

- cluster-proxy
- managed-serviceaccount
- work-manager

When your {mce-short} is imported into {product-title-short}, {product-title-short} enables the same set of addons to manage the {mce-short}. 

Install those addons in a different {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` addons while  {product-title-short} manages {mce-short} at the same time. Complete the following procedure:

. Log in to your {product-title-short} with the CLI.

. Create the addonsDeploymentConfig` resource to specify a different addon installation namespace. See the following example where `agentInstallNamespace` points to `open-cluster-management-agent-addons-discovery`:

+

> [source,yaml]

Reviewer 2: [source,bash]

> ----

Reviewer 2: ```
cat <<EOF | oc apply -f -
apiVersion: addons.open-cluster-management.io/v1alpha1
kind: addonsDeploymentConfig
metadata:
  name: addons-ns-config
  namespace: multicluster-engine
spec:
  agentInstallNamespace: open-cluster-management-agent-addons-discovery
EOF
```

apiVersion: addons.open-cluster-management.io/v1alpha1
kind: addonsDeploymentConfig
metadata:
  name: addons-ns-config
  namespace: multicluster-engine
spec:
  agentInstallNamespace: open-cluster-management-agent-addons-discovery
----

> . Update the existing `ClusterManagementaddons` resources for the addons so that the addonsDeploymentConfig` resource have the `agentInstallNamespace` namespace that you created.

Reviewer 2: Update the existing `ClusterManagementaddons` resources for the addons so that the addons are installed in `open-cluster-management-agent-addons-discovery` namespace specified in the `addonsDeploymentConfig` resource you created.

+

> See the following `ClusterManagementaddons` example resource for the `work-manager` addon:

Reviewer 2: Remove line 101 ~ 120.

Author: @rokej the step does not match the YAML so we have to tell the user what the YAML example is for. I can play with it but cannot remove it.

+
[source,yaml]
----
apiVersion: addons.open-cluster-management.io/v1alpha1
kind: ClusterManagementaddons
metadata:
  name: work-manager
spec:
  addonsMeta:
    displayName: work-manager
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
    type: Placements
----

> . Add the `addonsDeploymentConfig` to the `ClusterManagementaddons`. See the following example:

Reviewer 2: Step 3.a (a sub-step of line 98)

+
[source,yaml]
----
apiVersion: addons.open-cluster-management.io/v1alpha1
kind: ClusterManagementaddons
metadata:
  name: work-manager
spec:
  addonsMeta:
    displayName: work-manager
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io
        name: addons-ns-config
        namespace: multicluster-engine
        resource: addonsdeploymentconfigs
    type: Placements
----

> . Add the `addonsDeploymentConfig` to the `managed-serviceaccount`. See the following example:

Reviewer 2: Step 3.b (a sub-step of line 98)

+
[source,yaml]
----
apiVersion: addons.open-cluster-management.io/v1alpha1
kind: ClusterManagementaddons
metadata:
  name: managed-serviceaccount
spec:
  addonsMeta:
    displayName: managed-serviceaccount
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io
        name: addons-ns-config
        namespace: multicluster-engine
        resource: addonsdeploymentconfigs
    type: Placements
----

> . Add the `addonsDeploymentConfig` to the `cluster-proxy` addons. See the following example:

Reviewer 2: Step 3.c (a sub-step of line 98)

+
[source,yaml]
----
apiVersion: addons.open-cluster-management.io/v1alpha1
kind: ClusterManagementaddons
metadata:
  name: cluster-proxy
spec:
  addonsMeta:
    displayName: cluster-proxy
  installStrategy:
    placements:
    - name: global
      namespace: open-cluster-management-global-set
      rolloutStrategy:
        type: All
      configs:
      - group: addons.open-cluster-management.io
        name: addons-ns-config
        namespace: multicluster-engine
        resource: addonsdeploymentconfigs
    type: Placements
----

> The addons for the {product-title-short} `local-cluster` and all other managed clusters are re-installed into the namespace that you specified. 

Reviewer 2: The addons for the {product-title-short} `local-cluster` are re-installed into the namespace that you specified.

. Run the following command to verify:

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-addons-discovery
----

. See the following output:

+
[source,bash]
----
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h
klusterlet-addons-workmgr             1/1     1            1           24h
managed-serviceaccount-addons-agent   1/1     1            1           24h
----

[#create-klusterletconfig]
=== Creating a _KlusterletConfig_ resource

> 

Reviewer 2: Add this at the beginning of this section.

{mce-short} has a local-cluster, which is a hub cluster that is managed. A klusterlet resource named `klusterlet` is created for this local-cluster.

When your {mce-short} is imported into {product-title-short}, {product-title-short} installs the klusterlet with the same name `klusterlet` to manage the {mce-short} which will conflict with MCE's klusterlet for its local-cluster.

Install those addons in a different {mce-short} namespace so that the {mce-short} can self-manage with the `local-cluster` addons while  {product-title-short} manages {mce-short} at the same time. Complete the following proce

> When a `ManagedCluster` references the `KlusterletConfig` resource, the managed cluster `klusterlet` is installed in the namespace that you specified in the `KlusterletConfig`. 

Reviewer 2: Remove this line.

> Create a `KlusterletConfig` resource that is used by `ManagedCluster` resources to import {mce-short} clusters. 

Reviewer 2: Create a `KlusterletConfig` resource that is used by `ManagedCluster` resources to import {mce-short} clusters so that the klusterlet is installed with a different name to avoid the conflict.

> You can import the {product-title-short} klusterlet to be installed in a different namespace than the {mce-short} klusterlet for the {mce-short} `local-cluster`.

Reviewer 2: Remove this line.

> . Create a `KlusterletConfig` using the following example:

Reviewer 2: Create a `KlusterletConfig` using the following example. When this `KlusterletConfig` is referenced in a `ManagedCluster`, the value in the `spec.installMode.noOperator.postfix`  field is used as a suffix to the klusterlet name. e.g. `klusterlet-mce-import`

+
[source,yaml]

> ----

Reviewer 2: ```
cat <<EOF | oc apply -f -
kind: KlusterletConfig
apiVersion: config.open-cluster-management.io/v1alpha1
metadata:
  name: mce-import-klusterlet-config
spec:
  installMode:
    type: noOperator
    noOperator:
       postfix: mce-import
EOF
```

----

* For your `hypershift-addons-deploy-config`, run the following command:

+
[source,bash]
----
oc label addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine cluster.open-cluster-management.io/backup=true
----

* For your `work-manager`, run the following command:

+
[source,bash]
----
oc label clustermanagementaddons work-manager cluster.open-cluster-management.io/backup=true
----

* For your `cluster-proxy `, run the following command:

+
[source,bash]
----
oc label clustermanagementaddons cluster-proxy cluster.open-cluster-management.io/backup=true
----

* For your `managed-serviceaccount`, run the following command:

+
[source,bash]
----
oc label clustermanagementaddons managed-serviceaccount cluster.open-cluster-management.io/backup=true
----

* For your `mce-import-klusterlet-config`, run the following command:

+
[source,bash]
----
oc label KlusterletConfig mce-import-klusterlet-config cluster.open-cluster-management.io/backup=true
----

[#import]
== Importing {mce-short} manually

. From your {product-title-short} cluster, create a `ManagedCluster` resource manually to import an {mce-short} cluster. 

+

> [source,yaml]

Reviewer 2: [source,bash]

```
cat <<EOF | oc apply -f -
apiVersion: cluster.open-cluster-management.io/v1
kind: ManagedCluster
metadata:
  annotations:
    agent.open-cluster-management.io/klusterlet-config: mce-import-klusterlet-config <1>
  name: mce-a <2>
spec:
  hubAcceptsClient: true
  leaseDurationSeconds: 60
EOF
```

----
apiVersion: cluster.open-cluster-management.io/v1
kind: ManagedCluster
metadata:
  annotations:
    agent.open-cluster-management.io/klusterlet-config: mce-import-klusterlet-config <1>
  name: mce-a <2>
spec:
  hubAcceptsClient: true
  leaseDurationSeconds: 60
----

> <1> The `mce-import-klusterlet-config` annotation references the `KlusterletConfig` resource that you created in the previous step to install the {product-title-short} klusterlet into a different namespace in {mce-short}.

Reviewer 2: <1> The `mce-import-klusterlet-config` annotation references the `KlusterletConfig` resource that you created in the previous step to install the {product-title-short} klusterlet with a different name in {mce-short}.

<2> The example imports an {mce-short} managed cluster named `mce-a`.

> . Create a secret that references the `kubeconfig`` file of the cluster. Go to xref:../cluster_lifecycle/import_cli.adoc#importing-clusters-auto-import-secret [Importing a cluster by using the auto import secret] to add the auto import secret to complete the {mce-short} auto-import process. 

Reviewer 2: Create the `auto-import-secret` secret that references the `kubeconfig` of the {mce-short} cluster.

----

. Run the following command to disable metrics and to disable the hypershift operator management:

+
[source,bash]
----
oc patch addonsdeploymentconfig hypershift-addons-deploy-config -n multicluster-engine --type=merge -p '{"spec":{"customizedVariables":[{"name":"disableMetrics","value": "true"},{"name":"disableHOManagement","value": "true"}]}}'
----

. Run the following command to enable the `hypershift-addon` for {mce-short}:

+
[source,bash]
----
clusteradm addons enable --names hypershift-addon --clusters <managed-cluster-names>
----

. You can get the {mce-short} managed cluster names by running the following command in {product-title-short}.

+
[source,bash]
----
oc get managedcluster
----

. Log into {mce-short} clusters and verify that the hypershift addon is installed in the namespace that you specified. Run the following command:

+
[source,bash]
----
oc get deployment -n open-cluster-management-agent-addons-discovery
----
 
See the following example output that lists the addons:

+
[source,bash]

> ----

Reviewer 2: ```
NAME                                 READY   UP-TO-DATE   AVAILABLE   AGE
cluster-proxy-proxy-agent            1/1     1            1           24h
klusterlet-addons-workmgr           1/1     1            1           24h
hypershift-addons-agent             1/1     1            1           24h
managed-serviceaccount-addons-agent 1/1     1            1           24h
```
